File: Exercism/javascript/difference-of-squares/difference-of-squares.js
'use strict'

export class Squares {
  constructor (to) {
    const { numbers, squared, sum, value, compose } = this
    this._sumOfSquares = compose(
      numbers,
      squared,
      sum,
      value
    )(to)
    this._squareOfSum = compose(
      numbers,
      sum,
      squared,
      value
    )(to)
  }

  get sumOfSquares () {
    return this._sumOfSquares
  }

  get squareOfSum () {
    return this._squareOfSum
  }

  get difference () {
    return this._squareOfSum - this._sumOfSquares
  }

  * numbers (till) {
    for (let i = 1; i < till + 1; i += 1) {
      yield i
    }
  }

  * squared (iterable) {
    for (const n of iterable) {
      yield n ** 2
    }
  }

  * sum (iterable) {
    let sum = 0
    for (const n of iterable) {
      sum += n
    }
    yield sum
  }

  value (iterable) {
    const { value } = iterable.next()
    return value
  }

  compose (...fns) { 
    return (value) => {
      for (const fn of fns) {
        value = fn(value)
      }
      return value
    }
  }
}

File: Exercism/python/difference-of-squares/difference_of_squares.py
def square_of_sum(number: int) -> int:
    """Calc square of sum of numbers"""
    return sum(range(number + 1)) ** 2


def sum_of_squares(number: int) -> int:
    """Calc sum of squares of numbers"""
    return sum(map(lambda n: n ** 2, range(number + 1)))


def difference_of_squares(number: int) -> int:
    """Calc diff between sum of squares and square of sum"""
    return abs(sum_of_squares(number) - square_of_sum(number))
